glib/src/unichar.rs

```diff
@@ -0,0 +1,239 @@
+    }
+
+    fn is_punctuation(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_ispunct(self as _)) }
+    }
+
+    fn is_titlecase(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_istitle(self as _)) }
+    }
+
+    fn is_defined(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_isdefined(self as _)) }
+    }
+
+    fn is_wide(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_iswide(self as _)) }
+    }
+
+    fn is_wide_cjk(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_iswide_cjk(self as _)) }
+    }
+
+    fn is_zero_width(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_iszerowidth(self as _)) }
+    }
+
+    fn to_titlecase(self) -> char {
+        from_unichar(unsafe { ffi::g_unichar_totitle(self as _) })
+    }
+
+    fn mirror_char(self) -> Option<char> {
+        let mut mirrored = 0u32;
+        let res =
+            unsafe { bool::from_glib(ffi::g_unichar_get_mirror_char(self as _, &mut mirrored)) };
+        if res {
+            Some(from_unichar(mirrored))
+        } else {
+            None
+        }
+    }
+
+    fn fully_decompose(self, decomposition_kind: DecompositionKind) -> Vec<char> {
+        let buffer_len = ffi::G_UNICHAR_MAX_DECOMPOSITION_LENGTH as usize;
+        let mut buffer = vec![char::from(0); buffer_len];
+        let decomposition_length = unsafe {
+            ffi::g_unichar_fully_decompose(
+                self as _,
+                // NOTE: This assumes that glib will only ever write valid unicode codepoints to the
+                //       buffer.
+                decomposition_kind.into_bool().into_glib(),
+                buffer.as_mut_ptr().cast(),
+                buffer_len,
+            )
+        };
+        debug_assert!(decomposition_length <= buffer_len);
+        unsafe { buffer.set_len(decomposition_length) };
+        buffer
+    }
+
+    fn decompose(self) -> DecompositionResult {
+        let mut a = 0;
+        let mut b = 0;
+        let res = unsafe { bool::from_glib(ffi::g_unichar_decompose(self as _, &mut a, &mut b)) };
+
+        if res {
+            if b == 0 {
+                DecompositionResult::Singleton(from_unichar(a))
+            } else {
+                DecompositionResult::Pair(from_unichar(a), from_unichar(b))
+            }
+        } else {
+            DecompositionResult::NoDecomposition
+        }
+    }
+
+    fn compose(a: char, b: char) -> Option<char> {
+        let mut out = 0;
+        let res = unsafe { bool::from_glib(ffi::g_unichar_compose(a as _, b as _, &mut out)) };
+
+        if res {
+            Some(from_unichar(out))
+        } else {
+            None
+        }
+    }
+}
+
+fn from_unichar(unichar: u32) -> char {
+    // It's a glib bug if they give us an invalid unicode codepoint from one of their manipulation
+    // functions. It's safest if we just panic in that case
+    char::from_u32(unichar).unwrap()
```

Comment:
That's just what `char::from_glib()` is already doing or not?

Reply:
That doesn't seem to exist

Reply:
there is  char::try_from_glib though

Reply:
I don't think so? I can't find an impl anyway: https://github.com/search?q=repo%3Agtk-rs%2Fgtk-rs-core%20%22impl%20TryFromGlib%22&type=code

Reply:
Oh it seems to use TryFrom instead, see https://github.com/gtk-rs/gtk4-rs/blob/db7af24294c706136f3bcba74caf284957b9b9c2/gtk4/src/entry.rs#L22

Reply:
hm I guess I could switch to using TryFrom, the main point of this function is to (a) have the comment in one place (b) not repeat the actual expression to go from g_unichar -> char everywhere it's needed

I guess I could mark it as `#[inline]` and `#[track_caller]` though?

Reply:
Sounds like a good plan to me, but i would wait for @sdroege's opinion on this one :)

Reply:
Sounds like a plan